apps/arweave/src/ar_http_iface_middleware.erl

```diff
@@ -2358,6 +2321,7 @@ handle_block_announcement(#block_announcement{ indep_hash = H, previous_block =
 		solution_hash = SolutionH }, Req) ->
 	case ar_ignore_registry:member(H) of
 		true ->
+			ar_block_cache:update_timestamp(block_cache, H, erlang:timestamp()),
```

Comment:
Why do we update the timestamp here? We've seen the block before so shouldn't we have the relevant timestamp assigned already?

Reply:
Mostly to handle the fact that we only want to record the timestamp the first time the block is received from a peer so that we don't record the timestamp when the block was first mined. So in the event that this node mined a block it'll be added to the ignore registry even though the received timestamp hasn't been set. Instead we always check the timestamp whenever we receive an announcement or post, and update it if it hasn't already been set.

Reply:
since we do an ETS lookup by a specific key, I'm assuming the performance impact of the extra check whenever we handle a post or announcement is minimal . But perhaps that's a bad assumption?

Reply:
Are there any other places where we call update_timestamp except the two I see here under ar_ignore_registry:member(H) == true? Does it mean we only set the timestamp the second time we receive the block?

Reply:
We also set the timestamp the first time a block is pre-validated.

The value gets set in the Block here: https://github.com/ArweaveTeam/arweave/blob/master/apps/arweave/src/ar_block_pre_validator.erl#L49

And the block gets cached when the `{block, {new` event is handled, sent here: https://github.com/ArweaveTeam/arweave/blob/master/apps/arweave/src/ar_block_pre_validator.erl#L746

Reply:
ugh, I suppose I also need to handle the first time that a node fetches a block it had previously mined: https://github.com/ArweaveTeam/arweave/blob/master/apps/arweave/src/ar_poller.erl#L146

Reply:
Basically: it's the handling of self-mined blocks that causes all the trouble since we *don't* want the timestamp to be set to the momen the block was mined since that will make it possible to determine which nodes mined which blocks. And near as I can tell there are 3 places were we receive a block from a peer and need to check the timestamp:
- block announcement
- POST block
- GET block via polling
-

Reply:
Atually never mind, I don't think polling is relevant for this case. If the node has already mined the block it won't poll it from a peer, and if it hasn't mined the block then the timestamp will be handled via pre-validation when it polls it.

Reply:
So yeah, as far as I can tell:

Previous behavior:
1. receive_timestamp is *not* set when a node mines a block
2. receive_timestamp is set when a newly received block passes pre-validation

Desired behavior:
1. always set receive_timestamp the first time a node receives a block from a peer
2. To do this we need to handle the case where a block exists in a node's block_cache because it was the one who mined it. In this case we need to update the receive_timestamp when it receives the block either via `POST /block` or `POST /block_announcement` even though that block make already exist in the ar_ignore_registry

Let me know if you I've missed something or you think it should be implemented differently

Reply:
Ah, yes, I remember now, thank you for the detailed explanation!